include/swift/AST/TypeRepr.h

```diff
@@ -406,6 +406,13 @@ class GenericIdentTypeRepr final
     return {getTrailingObjects<TypeRepr*>(),
             Bits.GenericIdentTypeRepr.NumGenericArgs};
   }
+
+  bool argsAreProtocols(DeclContext *dc) {
```

Comment:
It looks like this predicate controls if we visit the generic arguments of a TypeRepr, but does it work if you have something like this? `Foo<Foo<P>>` where `P` is a protocol? The generic argument of `Foo<Foo<P>>` is `Foo<P>` which won't answer true to isProtocol() right? I think it would be simpler to just visit all children unconditionally.

Also, I suggest renaming isProtocol() to isProtocolOrProtocolComposition() or isConstraintTypeRepr() or something. It's not just protocols that become implicit generic parameters, it's protocol compositions

Reply:
updated 👍